### PR TITLE
Fix false-positive of `use-maxsplit-arg` when index is incremented inside a loop

### LINKED ISSUE #4664: `use-maxsplit-arg` false positive when index is incremented inside a loop
### Steps to reproduce

Given a file `a.py`:
```python
source = 'A.B.C.D.E.F.G'
other_iterable = range(5)
i = 0
for j in other_iterable:
    print(source.split('.')[i])
    i += 1
```

### Current behavior

Result of `pylint a.py`:
```
split.py:5: [C0207(use-maxsplit-arg), ] Use source.split('.', maxsplit=1)[0] instead
```

### Expected behavior
no messages, since maxsplit=1 produces a different result.

### pylint --version output
```
pylint 2.9.3
astroid 2.6.2
Python 3.8.10 (v3.8.10:3d8993a744, May  3 2021, 08:55:58) 
```

### PR DESCRIPTION
**Pull request by yushao2**
## Steps

- [x] Add a ChangeLog entry describing what your PR does.
- [x] Write a good description on what the PR does.

## Description

Handle case where variable subscripts are mutated within a loop, see #4664

Minor refactoring to if/else logic as there were linting issues where there's too much if/else nesting

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Related Issue
Closes #4664

### CODE REVIEW

File: pylint/checkers/refactoring/recommendation_checker.py [hunk 1]
```diff
@@ -78,38 +78,58 @@ def _check_use_maxsplit_arg(self, node: astroid.Call) -> None:
...
-                        node.func.as_string().rsplit(fn_name, maxsplit=1)[0]
-                        + new_fn
-                        + f"({node.args[0].as_string()}, maxsplit=1)[{subscript_value}]"
-                    )
-                    self.add_message("use-maxsplit-arg", node=node, args=(new_name,))
+            # Check for cases where variable (Name) subscripts may be mutated within a loop
+            if isinstance(node.parent.slice, astroid.Name):
+                # Check if loop present within the scope of the node
+                scope = node.scope()
+                for loop_node in scope.nodes_of_class((astroid.For, astroid.While)):
```
Comment (cdce8p) [inline pylint/checkers/refactoring/recommendation_checker.py:110]: Could you add type hints for `loop_node` and `assignment_node` below (with `typing.cast`).

I should probably add types to `NodeNG.nodes_of_class` soon, so we don't need to anymore.

File: tests/functional/u/use/use_maxsplit_arg.py [hunk 2]
```diff
@@ -82,3 +82,12 @@ class Bar():
 a = "1,2,3".split('\n')[0]  # [use-maxsplit-arg]
 a = "1,2,3".split('split')[-1]  # [use-maxsplit-arg]
 a = "1,2,3".rsplit('rsplit')[0]  # [use-maxsplit-arg]
+
+# Test cases for false-positive reported in #4664
+# https://github.com/PyCQA/pylint/issues/4664
+source = 'A.B.C.D.E.F.G'
+other_iterable = range(5)
+i = 0
+for j in other_iterable:
```
Comment (cdce8p) [inline tests/functional/u/use/use_maxsplit_arg.py:91]:
```suggestion
for j in range(5):
```
I would simplify it a bit

File: tests/functional/u/use/use_maxsplit_arg.txt [hunk 3]
```diff
@@ -1,21 +1,21 @@
...
-use-maxsplit-arg:66:10::Use s.split(' ', maxsplit=1)[0] instead
-use-maxsplit-arg:67:10::Use s.rsplit(' ', maxsplit=1)[-1] instead
-use-maxsplit-arg:76:6::Use Bar.split.split(',', maxsplit=1)[0] instead
-use-maxsplit-arg:77:6::Use Bar.split.rsplit(',', maxsplit=1)[-1] instead
-use-maxsplit-arg:78:6::Use Bar.split.split(',', maxsplit=1)[0] instead
-use-maxsplit-arg:79:6::Use Bar.split.rsplit(',', maxsplit=1)[-1] instead
-use-maxsplit-arg:82:4::Use '1,2,3'.split('\n', maxsplit=1)[0] instead
-use-maxsplit-arg:83:4::Use '1,2,3'.rsplit('split', maxsplit=1)[-1] instead
-use-maxsplit-arg:84:4::Use '1,2,3'.split('rsplit', maxsplit=1)[0] instead
+use-maxsplit-arg:5:12::Use '1,2,3'.split(',', maxsplit=1)[0] instead:HIGH
```
Comment (cdce8p) [inline tests/functional/u/use/use_maxsplit_arg.txt:1]: Just my opinion, but I'm not a big fan of adding `:HIGH` and creating larger diffs than necessary. That makes it quite difficult to review.

**Review by cdce8p — APPROVED** [review]
Looks good! Added a 2-3 small notes. Feel free to merge it once addressed.

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM 😄 Only small thing is that the refactor could have been done in another MR first so the real changes are easier to review :)

**Pull request merged**